### PR TITLE
V4.0 Lasrge Tx Support

### PR DESCRIPTION
**Pull request by aviviadi**
http://issues.hibernatingrhinos.com/issue/RavenDB-4806

### CODE REVIEW

File: src/Raven.Server/Config/Categories/StorageConfiguration.cs [hunk 1]
```diff
@@ -78,6 +78,12 @@ public class StorageConfiguration : ConfigurationCategory
         [LegacyConfigurationEntry("Raven/TransactionsModeDuration")]
         public int TransactionsModeDuration { get; set; }
 
+        [Description("Should voron support large transactions")]
+        [DefaultValue(false)]
+        [ConfigurationEntry("Raven/Storage/SupportLargeTransactions")]
+        [LegacyConfigurationEntry("Raven/SupportLargeTransactions")]
+        public bool SupportLargeTransactions { get; set; }
```
Comment (ayende) [inline src/Raven.Server/Config/Categories/StorageConfiguration.cs]: Should remove this, we are always supporting large transactions

File: src/Sparrow/Compression/LZ4.cs [hunk 2]
```diff
@@ -79,6 +79,35 @@ protected struct LZ4_stream_t_internal
...
         }
 
+        public long Encode64LongBuffer(
+            byte* input,
+            byte* output,
+            long inputLength,
+            long outputLength,
+            int acceleration = ACCELERATION_DEFAULT)
+        {
+            if (inputLength < int.MaxValue && outputLength < int.MaxValue)
```
Comment (ayende) [inline src/Sparrow/Compression/LZ4.cs]: There is an error here, LZ4 can't actually handle 2GB size, it is something a bit less than that, and you need to take that into account.

File: src/Sparrow/Compression/LZ4.cs [hunk 3]
```diff
@@ -79,6 +79,35 @@ protected struct LZ4_stream_t_internal
...
+            long totalOutputSize = 0;
+            long pos = 0;
+            while (pos < inputLength)
+            {
+                int partInputLength = int.MaxValue;
+                if (pos + partInputLength > inputLength)
+                    partInputLength = (int)(Int32.MaxValue - pos);
+
+                int partOutputLength = int.MaxValue;
+                totalOutputSize += Encode64(input, output, partInputLength, partOutputLength, acceleration);
```
Comment (ayende) [inline src/Sparrow/Compression/LZ4.cs]: In here, you are always compressing the first part of the data, and never the latter part.
The `input` and `output` values aren't change, and that is what it is using.

File: src/Sparrow/Compression/LZ4.cs (hunk 2 above)
Comment (ayende) [reply src/Sparrow/Compression/LZ4.cs]: At any rate, we are probably better at compressing at 1GB size, to ensure that we are away from the limit.

File: src/Sparrow/Compression/LZ4.cs [hunk 4]
```diff
@@ -505,6 +534,35 @@ private static TResult ThrowException<TResult>(Exception e)
...
         private const ulong prime5bytes = 889523592379UL;
 
+        public static long Decode64LongBuffers(
+            byte* input,
+            long inputLength,
+            byte* output,
+            long outputLength,
+            bool knownOutputLength)
+        {
+            if (inputLength < int.MaxValue && outputLength < int.MaxValue)
```
Comment (ayende) [inline src/Sparrow/Compression/LZ4.cs]: Same size issue here.

File: src/Sparrow/Compression/LZ4.cs [hunk 5]
```diff
@@ -505,6 +534,35 @@ private static TResult ThrowException<TResult>(Exception e)
...
+            long totalOutputSize = 0;
+            long pos = 0;
+            while (pos < inputLength)
+            {
+                int partInputLength = int.MaxValue;
+                if (pos + partInputLength > inputLength)
+                    partInputLength = (int) (Int32.MaxValue - pos);
+
+                int partOutputLength = int.MaxValue;
+                totalOutputSize += Decode64(input, partInputLength, output, partOutputLength, false);
```
Comment (ayende) [inline src/Sparrow/Compression/LZ4.cs]: Same decoding issue here, you are not moving the input & output here.

File: src/Voron/Impl/Backup/IncrementalBackup.cs [hunk 6]
```diff
@@ -274,7 +274,7 @@ public void Restore(string outPath, IEnumerable<string> backupPaths)
                         return;
                     foreach (var dir in package.Entries.GroupBy(entry => Path.GetDirectoryName(entry.FullName)))
                     {
-                        using (var options = StorageEnvironmentOptions.ForPath(Path.Combine(outPath, dir.Key)))
+                        using (var options = StorageEnvironmentOptions.ForPath(Path.Combine(outPath, dir.Key), false)) // TODO : should we support here large tx ?
```
Comment (ayende) [inline src/Voron/Impl/Backup/IncrementalBackup.cs]: Yes

File: src/Voron/Impl/Journal/JournalFile.cs [hunk 7]
```diff
@@ -149,7 +149,8 @@ public void Write(LowLevelTransaction tx, CompressedPagesResult pages, LazyTrans
 
             if (tx.IsLazyTransaction == false && (lazyTransactionScratch == null || lazyTransactionScratch.HasDataInBuffer() == false))
             {
-                _journalWriter.WritePages(position, pages.Base, pages.NumberOfPages);
+                Debug.Assert(pages.NumberOfPages < int.MaxValue);
```
Comment (ayende) [inline src/Voron/Impl/Journal/JournalFile.cs]: This is not enough here, we need to actually handle this scenario

Comment (ayende) [reply src/Voron/Impl/Journal/JournalFile.cs]: And in general, if the transaction size is bigger than 2GB, it isn't lazy.
Also, you are comparing the number of pages to 2GB, but that isn't the actual size.

File: src/Voron/Impl/Journal/JournalReader.cs [hunk 8]
```diff
@@ -126,9 +126,15 @@ public bool ReadOneTransactionToDataFile(StorageEnvironmentOptions options, bool
...
         }
 
-        internal static int GetNumberOfPagesFromSize(StorageEnvironmentOptions options, int size)
+        internal static int GetNumberOfPagesFromSize(StorageEnvironmentOptions options, long size)
         {
-            return (size / options.PageSize) + (size % options.PageSize == 0 ? 0 : 1);
+            var lastPage = (size % options.PageSize == 0 ? 0 : 1);
+            var result = (size/options.PageSize) + lastPage;
+            if (result > int.MaxValue)
+                throw new InvalidDataException(
```
Comment (ayende) [inline src/Voron/Impl/Journal/JournalReader.cs]: Move the throw to a dedicated method.

File: src/Voron/Impl/Journal/JournalReader.cs [hunk 9]
```diff
@@ -126,9 +126,15 @@ public bool ReadOneTransactionToDataFile(StorageEnvironmentOptions options, bool
...
+                    $"GetNumberOfOverflowPages returned number exceeding int.MaxValue : {result}");
+            int numOfPages = unchecked((int)result);
```
Comment (ayende) [inline src/Voron/Impl/Journal/JournalReader.cs]: You don't need the unchecked here.

File: src/Voron/Impl/Journal/LazyTransactionBuffer.cs [hunk 10]
```diff
@@ -33,9 +33,11 @@ public void AddToBuffer(long position, CompressedPagesResult pages, int uncompre
             {
                 _firstPositionInJournalFile = position; // first lazy tx saves position to all lazy tx that comes afterwards
             }
-            _lazyTransactionPager.WriteDirect(pages.Base, _lastUsedPage, pages.NumberOfPages);
 
-            _lastUsedPage += pages.NumberOfPages;
+            Debug.Assert(pages.NumberOfPages < int.MaxValue);
```
Comment (ayende) [inline src/Voron/Impl/Journal/LazyTransactionBuffer.cs]: Wrong calculation here, and assert doesn't cut it.

File: src/Voron/Impl/Journal/WriteAheadJournal.cs [hunk 11]
```diff
@@ -845,7 +845,7 @@ public void WriteToJournal(LowLevelTransaction tx, int pageCount)
             if (CurrentFile == null || CurrentFile.AvailablePages < pages.NumberOfPages)
             {
                 _lazyTransactionBuffer?.WriteBufferToFile(CurrentFile, tx);
-                CurrentFile = NextFile(pages.NumberOfPages);
+                CurrentFile = NextFile((int)pages.NumberOfPages);
```
Comment (ayende) [inline src/Voron/Impl/Journal/WriteAheadJournal.cs]: There seems to be a repeating pattern of mixing the number of pages and the actual size.
int.MaxValue \* 4KB = 16TB.

File: src/Voron/Impl/Journal/WriteAheadJournal.cs [hunk 12]
```diff
@@ -867,15 +867,21 @@ private CompressedPagesResult PrepreToWriteToJournal(LowLevelTransaction tx, Abs
...
             // We want to include the Transaction Header straight into the compression buffer.
             var sizeOfPagesHeader = numberOfPages * sizeof(TransactionHeaderPageInfo);
-            var maxSizeRequiringCompression = pageCountIncludingAllOverflowPages * pageSize + sizeOfPagesHeader
+            long maxSizeRequiringCompression = ((long)pageCountIncludingAllOverflowPages * (long)pageSize) + sizeOfPagesHeader
                  + numberOfPages * sizeof(long);
-            var outputBufferSize = LZ4.MaximumOutputLength(maxSizeRequiringCompression);
+
+            long outputBufferSize;
+            if (maxSizeRequiringCompression > int.MaxValue)
+                outputBufferSize = maxSizeRequiringCompression + (maxSizeRequiringCompression / 255) + 16;
```
Comment (ayende) [inline src/Voron/Impl/Journal/WriteAheadJournal.cs]: Explain the math here, please.

File: src/Voron/Impl/Paging/AbstractPagerExtensions.cs [hunk 13]
```diff
@@ -45,10 +46,15 @@ public static int GetNumberOfPages(this AbstractPager pager, Page page)
             return page.IsOverflow ? pager.GetNumberOfOverflowPages(page.OverflowSize) : 1;
         }
 
-        public static int GetNumberOfOverflowPages(this AbstractPager pager, int overflowSize)
+        public static int GetNumberOfOverflowPages(this AbstractPager pager, long overflowSize)
         {
             overflowSize += Constants.TreePageHeaderSize;
-            return (overflowSize/pager.PageSize) + (overflowSize%pager.PageSize == 0 ? 0 : 1);
+            long result = (overflowSize/pager.PageSize) + (overflowSize%pager.PageSize == 0 ? 0 : 1);
+            if (result > int.MaxValue)
```
Comment (ayende) [inline src/Voron/Impl/Paging/AbstractPagerExtensions.cs]: Same pages vs size error.

File: src/Voron/Impl/Scratch/ScratchBufferPool.cs [hunk 14]
```diff
@@ -206,8 +206,17 @@ public PageFromScratchBuffer Allocate(LowLevelTransaction tx, int numberOfPages)
 
                     createNextFile = true;
                 }
+                else if (_options.SupportLargeTransactions)
```
Comment (ayende) [inline src/Voron/Impl/Scratch/ScratchBufferPool.cs]: This should be the default, but we need to have a discussion on this to make sure we have a coherent strategy here.

File: src/Voron/StorageEnvironmentOptions.cs [hunk 15]
```diff
@@ -95,6 +95,8 @@ public int PageSize
 
         public long? MaxStorageSize { get; set; }
 
+        public bool SupportLargeTransactions { get; set; }
```
Comment (ayende) [inline src/Voron/StorageEnvironmentOptions.cs]: Remove this, it is the deafult.

File: test/SlowTests/Voron/HugeTransactions.cs [hunk 16]
```diff
@@ -0,0 +1,120 @@
...
+        const string chars = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890 ";
+        public static Random Rand = new Random(123);
+
+        [Theory]
+        [InlineData(1)]
+        [InlineData(2)]
+        [InlineData(4)]
+        [InlineData(6)]
+        [InlineData(9)]
+        public unsafe void CanWriteBigTransactions(long transactionSizeInGb)
```
Comment (ayende) [inline test/SlowTests/Voron/HugeTransactions.cs:36]: Those are actually running concurrently, and it is really killing the test machine.
We can just have 2 & 6 GB tests, and make sure that they aren't running concurrently.

**Review by ayende — CHANGES_REQUESTED** [review]

File: src/Voron/Platform/Win32/Win32JournalWriter.cs [hunk 17]
```diff
@@ -55,11 +55,22 @@ public void WritePages(long position, byte* p, int numberOfPages)
             if (Disposed)
                 throw new ObjectDisposedException("Win32JournalWriter");
 
-            if (numberOfPages > _maxNumberOfPages)
-                throw new ArgumentException(
-                    "Cannot write " + numberOfPages + " pages in a single write, the size is too large",
-                    nameof(numberOfPages));
+            while (numberOfPages > _maxNumberOfPagesPerSingleWrite)
+            {
+                WriteFile(position, p, _maxNumberOfPagesPerSingleWrite);
```
Comment (aviviadi) [inline src/Voron/Platform/Win32/Win32JournalWriter.cs:60]: TODO : We need to make a single call or mark the entire section until we finish all the segments writes

File: src/Voron/Impl/Journal/WriteAheadJournal.cs [hunk 18]
```diff
@@ -909,14 +912,14 @@ private CompressedPagesResult PrepreToWriteToJournal(LowLevelTransaction tx, Abs
                 pagesInfo[pageSequencialNumber].DiffSize = _diffPage.IsDiff ? _diffPage.OutputSize : 0;
                 ++pageSequencialNumber;
             }
-            var totalSizeWritten = (int)(write - outputBuffer + sizeOfPagesHeader);
+            var totalSizeWritten = (write - outputBuffer) + sizeOfPagesHeader;
 
 
-            var fullTxBuffer = outputBuffer + pageCountIncludingAllOverflowPages * pageSize;
+            var fullTxBuffer = outputBuffer + (pageCountIncludingAllOverflowPages*(long)pageSize);
```
Comment (aviviadi) [inline src/Voron/Impl/Journal/WriteAheadJournal.cs:918]: This "little" cast was the buggy issue for large writes

**Pull request merged**